compiler-core/src/format.rs

```diff
@@ -1487,22 +1500,54 @@ fn list<'a>(elems: Document<'a>, tail: Option<Document<'a>>) -> Document<'a> {
 }
 
 fn printed_comments<'a, 'comments>(
-    comments: impl IntoIterator<Item = &'comments str>,
+    comments: impl IntoIterator<Item = Option<&'comments str>>,
+    trailing_newline: bool,
 ) -> Option<Document<'a>> {
     let mut comments = comments.into_iter().peekable();
     let _ = comments.peek()?;
-    Some(concat(Itertools::intersperse(
-        comments.map(|c| "//".to_doc().append(Document::String(c.to_string()))),
-        line(),
-    )))
+
+    let mut doc = Vec::new();
+    while let Some(c) = comments.next() {
+        if let Some(c) = c {
+            doc.push("//".to_doc().append(Document::String(c.to_string())));
+            match comments.peek() {
+                // next line is a comment
+                Some(Some(_)) => doc.push(line()),
+                // next line is empty
+                Some(None) => {
+                    let _ = comments.next();
+                    match comments.peek() {
+                        Some(_) => doc.push(lines(2)),
+                        None => {
+                            if trailing_newline {
+                                doc.push(force_break());
+                                doc.push(lines(2));
+                            }
+                        }
+                    }
+                }
+                None => {
+                    if trailing_newline {
+                        doc.push(force_break());
+                        doc.push(line());
+                    }
+                }
+            }
+        } else {
+            // there will never be consecutive empty lines (None values),
+            // and whenever we peek a None, we advance past it
+            unreachable!()
+        }
+    }
```

Comment:
This is a bit gnarly, but I wasn't able to simplify any further without introducing whitespace on empty lines. Perhaps someone more familiar with the formatting approach could recommend a better way?

Reply:
We could replace the `if let Some(c) = c {` with `c.expect("consecutive empty lines")`, that'd remove one level of indentation.